src/qibo/backends/numpy.py

```diff
@@ -83,811 +46,9 @@ def cast(self, x, dtype=None, copy: bool = False):
-            matrix_log = logm(self.to_numpy(matrix)) / float(np.log(base))
-            matrix_log = self.cast(matrix_log, dtype=matrix_log.dtype)
-
-            return matrix_log
-
-        # log = self.np.diag(self.np.log(eigenvalues) / float(np.log(base)))
-        log = self.np.log(eigenvalues) / float(np.log(base))
-        ud = self.np.transpose(np.conj(eigenvectors))
-
-        return (eigenvectors * log) @ ud
-
-    def calculate_matrix_power(
-        self,
-        matrix,
-        power: Union[float, int],
-        precision_singularity: float = 1e-14,
-    ):
-        if not isinstance(power, (float, int)):
-            raise_error(
-                TypeError,
-                f"``power`` must be either float or int, but it is type {type(power)}.",
-            )
-
-        if power < 0.0:
-            # negative powers of singular matrices via SVD
-            determinant = self.np.linalg.det(matrix)
-            if abs(determinant) < precision_singularity:
-                return _calculate_negative_power_singular_matrix(
-                    matrix, power, precision_singularity, self.np, self
-                )
-
-        return fractional_matrix_power(matrix, power)
-
-    def calculate_matrix_sqrt(self, matrix):
-        return self.calculate_matrix_power(matrix, power=0.5)
-
-    def calculate_singular_value_decomposition(self, matrix):
-        return self.np.linalg.svd(matrix)
-
-    def calculate_jacobian_matrix(
-        self, circuit, parameters=None, initial_state=None, return_complex: bool = True
-    ):
-        raise_error(
-            NotImplementedError,
-            "This method is only implemented in backends that allow automatic differentiation, "
-            + "e.g. ``PytorchBackend`` and ``TensorflowBackend``.",
-        )
-
-    def assert_allclose(self, value, target, rtol=1e-7, atol=0.0):
-        if isinstance(value, CircuitResult) or isinstance(value, QuantumState):
-            value = value.state()
-        if isinstance(target, CircuitResult) or isinstance(target, QuantumState):
-            target = target.state()
-        value = self.to_numpy(value)
-        target = self.to_numpy(target)
-        np.testing.assert_allclose(value, target, rtol=rtol, atol=atol)
-
-
-def _calculate_negative_power_singular_matrix(
-    matrix, power: Union[float, int], precision_singularity: float, engine, backend
-):
-    """Calculate negative power of singular matrix."""
-    U, S, Vh = backend.calculate_singular_value_decomposition(matrix)
-    # cast needed because of different dtypes in `torch`
-    S = backend.cast(S)
-    S_inv = engine.where(engine.abs(S) < precision_singularity, 0.0, S**power)
+        return self.engine.asarray(x, dtype=dtype, copy=copy if copy else None)
 
-    return engine.linalg.inv(Vh) @ backend.np.diag(S_inv) @ engine.linalg.inv(U)
+    def to_numpy(self, array):
+        if self.is_sparse(array):
+            return array.toarray()
+        return array
```

Comment:
Up to personal taste, but in these cases I favor the ternary `if`, which is not breaking the flow, but just defines a conditional value.

```suggestion
        return array.toarray() if self.is_sparse(array) else array
```